### PR TITLE
#791 - Survey Monkey new actions

### PR DESCRIPTION
**Pull request by GTFalcao**
# New Actions
- **List Surveys**
- **List Responses**
- **List Collectors**
- **Get Response Details**
- **Get Collector Details**

# Modified Action
- **Find Survey**
  - Changed name and description to **Get Survey Details** to maintain a clear pattern between the actions, but did not rename the actual component key `find-survey` - is this the correct approach?

# Modified Source
- **Custom Webhook Events**
  - Adjusted the code to add support for pagination with `_paginatedRequest` returning an array instead of a single request response. Previously, only the first 50 values (default SurveyMonkey response page size) would be received by `getSurveys` and `getCollectors`

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/survey_monkey/sources/custom-webhook-events/custom-webhook-events.mjs [hunk 1]
```diff
@@ -4,7 +4,7 @@ import common from "../common/base.mjs";
 export default {
   ...common,
   name: "New Custom webhook events",
-  version: "0.0.1",
+  version: "0.0.2",
   type: "source",
   key: "survey_monkey-custom-webhook-events",
   description: "Emmit new custom webhook event",
```
Comment (jcortes) [inline components/survey_monkey/sources/custom-webhook-events/custom-webhook-events.mjs:10]:
```suggestion
  description: "Emit new custom webhook event",
```

File: components/survey_monkey/survey_monkey.app.mjs [hunk 2]
```diff
@@ -9,16 +9,52 @@ export default {
...
           value: survey.id,
         }));
       },
     },
+    responseId: {
+      type: "string",
+      label: "Response",
+      description: `Select one of the above survey's **Responses** to get details of.
+      \\
+      Alternatively, you can provide a custom *Response ID*.`,
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:29]:
```suggestion
      description: `Select one of the above survey's **Responses** to get details of. Alternatively, you can provide a custom *Response ID*.`,
```

File: components/survey_monkey/survey_monkey.app.mjs [hunk 3]
```diff
@@ -9,16 +9,52 @@ export default {
...
+          value: response.id,
+        }));
+      },
+    },
+    collectorId: {
+      type: "string",
+      label: "Collector",
+      description: `Select one of the above survey's **Collectors** to get details of.
+      \\
+      Alternatively, you can provide a custom *Collector ID*.`,
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:46]:
```suggestion
      description: `Select one of the above survey's **Collectors** to get details of. Alternatively, you can provide a custom *Collector ID*.`,
```

File: components/survey_monkey/survey_monkey.app.mjs [hunk 4]
```diff
@@ -50,13 +86,7 @@ export default {
...
     async _makeRequest(customConfig) {
       const {
-        $,
-        url,
-        path,
-        method,
-        params,
-        data,
-        ...otherConfig
+        $, url, path, method, params, data, ...otherConfig
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:89]:
```suggestion
        $ = this, url, path, method, params, data, ...otherConfig
```
and then you can call `axios` like
```javascript
return axios($, config);
```

File: components/survey_monkey/survey_monkey.app.mjs [hunk 5]
```diff
@@ -9,16 +9,52 @@ export default {
     survey: {
       type: "string",
       label: "Survey",
-      description: "Survey where the action will be performed.",
+      description: `Select a **Survey** from the list.
+      \\
+      Alternatively, you can provide a custom *Survey ID*.`,
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:14]:
```suggestion
      description: `Select a **Survey** from the list. Alternatively, you can provide a custom *Survey ID*.`,
```

File: components/survey_monkey/survey_monkey.app.mjs [hunk 6]
```diff
@@ -73,25 +103,86 @@ export default {
...
+          ...params,
+          path: `${path}?page=${page}&per_page=${amountPerPage}`,
+        });
+
+        if (response.data) values.push(...response.data);
+        else throw new Error(response);
+      } while (page++ < Math.ceil(response.total / amountPerPage));
+
+      return values;
+    },
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:128]: To simplify a bit more taking advantage of `axios` properties:
```suggestion
    async _paginatedRequest(args = {}) {
      // https://api.surveymonkey.net/v3/docs?javascript#pagination
      // using default SurveyMonkey pagination - 50 per page
      const amountPerPage = 50;
      const values = [];

      let page = 1;
      let response;

      do {
        response = await this._makeRequest({
          params: {
            page,
            per_page: amountPerPage,
          },
          ...args,
        });

        if (response.data) {
          values.push(...response.data);
        } else {
          throw new Error(response);
        }
      } while (page++ < Math.ceil(response.total / amountPerPage));

      return values;
    },
```

File: components/survey_monkey/survey_monkey.app.mjs [hunk 7]
```diff
@@ -73,25 +103,86 @@ export default {
...
+    async getUserInfo({ $ }) {
+      return this._makeRequest({
+        method: "GET",
+        path: "/users/me",
+        $,
+      });
+    },
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:135]: You can simplify here in that way you don't have to set `$` symbol no each function. This same advice applies to all other functions with this approach.
```suggestion
    async getUserInfo(args = {}) {
      return this._makeRequest({
        method: "GET",
        path: "/users/me",
        ...args,
      });
    },
```

**Review by jcortes — CHANGES_REQUESTED** [review]
HI @GTFalcao I've just added a few request changes. Overall it's looking good so far. thanks

File: components/survey_monkey/sources/custom-webhook-events/custom-webhook-events.mjs (hunk 1 above)
Comment (GTFalcao) [reply components/survey_monkey/sources/custom-webhook-events/custom-webhook-events.mjs:10]: Great catch! Fixed that typo.

File: components/survey_monkey/survey_monkey.app.mjs (hunk 4 above)
Comment (GTFalcao) [reply components/survey_monkey/survey_monkey.app.mjs:89]: Agreed! Applied this change.

File: components/survey_monkey/survey_monkey.app.mjs (hunk 7 above)
Comment (GTFalcao) [reply components/survey_monkey/survey_monkey.app.mjs:135]: That makes sense, but most of the methods use some of the `args` to assemble the URL, for example:
`async getResponseDetails({
      surveyId, responseId, $,
    }) {
      return this._makeRequest({
        method: "GET",
        path: `/surveys/${surveyId}/responses/${responseId}`,
        $,
      });
    },`
In this case, since surveyId and responseId will need to be destructured from the `args`, I prefer to destructure all of them right away; this also immediately shows what the method uses.

Do you think this change should be applied only for `getUserInfo` and `getSurveys` which require no parameters other than $? Or do you think there is a better approach to keep all the methods with the same pattern?

File: components/survey_monkey/survey_monkey.app.mjs (hunk 6 above)
Comment (GTFalcao) [reply components/survey_monkey/survey_monkey.app.mjs:128]: Great suggestion! I overlooked the use of axios `params` completely. Changed to use them as per your suggestion.

File: components/survey_monkey/survey_monkey.app.mjs (hunk 2 above)
Comment (GTFalcao) [reply components/survey_monkey/survey_monkey.app.mjs:29]: This was done to provide a line break in the prop description - do you think it'd be best to remove that line break then?

Here's a visual comparison of the suggested change:
image

File: components/survey_monkey/survey_monkey.app.mjs (hunk 7 above)
Comment (jcortes) [reply components/survey_monkey/survey_monkey.app.mjs:140]: Yeah in theses case what you can do is
```javascript
    async getResponseDetails({
      surveyId, responseId, ...args
    } = {}) {
      return this._makeRequest({
        method: "GET",
        path: `/surveys/${surveyId}/responses/${responseId}`,
        ...args,
      });
    },
```
in that way you don't deal with `$` here either

File: components/survey_monkey/survey_monkey.app.mjs [hunk 8]
```diff
@@ -101,27 +101,32 @@ export default {
...
 
-      return axios($ || this, config);
+      return axios($, config);
     },
-    async _paginatedRequest(params) {
+    async _paginatedRequest(args = {}) {
       // https://api.surveymonkey.net/v3/docs?javascript#pagination
       // using default SurveyMonkey pagination - 50 per page
-      const amountPerPage = 50;
+      const amountPerPage = 1;
```
Comment (jcortes) [inline components/survey_monkey/survey_monkey.app.mjs:109]: Why did you change `50` to `1`?

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @GTFalcao I have one minor observation about pagination. Thanks

File: components/survey_monkey/survey_monkey.app.mjs (hunk 7 above)
Comment (GTFalcao) [reply components/survey_monkey/survey_monkey.app.mjs]: Sure! That works well and is even more clear about which args are explicitly needed for this method.

Agreed and applied!

File: components/survey_monkey/survey_monkey.app.mjs (hunk 8 above)
Comment (GTFalcao) [reply components/survey_monkey/survey_monkey.app.mjs]: That was a mistake - I had changed it for testing and should not have commited that change. Fixed this, thanks!

**Review by jcortes — COMMENTED** [review]
HI @GTFalcao LGTM! Ready for QA!

**Comment by DilanAthukorala** [comment]
Verification completed for newly added survey monkey actions and no issues observed.

**Review by michelle0927 — APPROVED** [review]
